src/nbl/video/CVulkanCommandBuffer.h

```diff
@@ -3,28 +3,435 @@
 
 #include "nbl/video/IGPUCommandBuffer.h"
 
+// Todo(achal): I think I a lot of them could be made forward declarations if I introduce
+// a CVulkanCommandBuffer.cpp
+#include "nbl/video/CVulkanBuffer.h"
+#include "nbl/video/CVulkanImage.h"
+#include "nbl/video/CVulkanComputePipeline.h"
+#include "nbl/video/CVulkanPipelineLayout.h"
+#include "nbl/video/CVulkanDescriptorSet.h"
+
 #include <volk.h>
 
-namespace nbl
-{
-namespace video
+namespace nbl::video
 {
+struct ArgumentReferenceSegment;
 
-class CVulkanCommandBuffer : public virtual IGPUCommandBuffer
+class CVulkanCommandBuffer : public IGPUCommandBuffer
 {
 public:
-    VkCommandBuffer getInternalObject() const { return m_cmdbuf; }
+    CVulkanCommandBuffer(core::smart_refctd_ptr<ILogicalDevice>&& logicalDevice, E_LEVEL level,
+        VkCommandBuffer _vkcmdbuf, IGPUCommandPool* commandPool)
+        : IGPUCommandBuffer(std::move(logicalDevice), level, commandPool), m_cmdbuf(_vkcmdbuf)
+    {
+        if (m_cmdpool->getAPIType() == EAT_VULKAN)
+        {
+            CVulkanCommandPool* vulkanCommandPool = static_cast<CVulkanCommandPool*>(m_cmdpool.get());
+            vulkanCommandPool->emplace_n(m_argListTail, nullptr, nullptr);
+            m_argListHead = m_argListTail;
+        }
+    }
+
+    ~CVulkanCommandBuffer()
+    {
+        freeSpaceInCmdPool();
+    }
+
+    // API needs to change, vkBeginCommandBuffer can fail
+    void begin(uint32_t recordingFlags) override
+    {
+        // Should we do manual state management? Wouldn't relying on validation errors be better?
+        // IGPUCommandBuffer::begin(recordingFlags);
+        
+        VkCommandBufferBeginInfo beginInfo = { VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO };
+        beginInfo.pNext = nullptr; // pNext must be NULL or a pointer to a valid instance of VkDeviceGroupCommandBufferBeginInfo
+        beginInfo.flags = static_cast<VkCommandBufferUsageFlags>(recordingFlags);
```

Comment:
can you just cast our engine enums by value to vulkan enums?

Reply:
I believe there should be converter functions for these things that needs back and forth conversion?

Even though they might be the same for now, They might differ when upgrading Vulkan to a new version.

Also be aware of flags (VkFlags) that can have multiple enum values OR'ed together and care must be taken when converting to/from NBL::asset enums.

Reply:
> Even though they might be the same for now, They might differ when upgrading Vulkan to a new version.

my idea to have a bunch of `static_asserts` that enforce that they're always directly castable.